FQDN: some small performance optimizations around logging

squeed opened the pull request:
As part of some basic benchmarking, I noticed that we were spending a lot of time in the DNS critical section enriching Hubble logs. Additionally, we were wasting CPU time re-generating information we already had!

So, there are 3 changes:

- Change the label representation of some internal types from []string to labels.LabelArray. We already have the LabelArray on hand, so we can save the allocation and string conversion.
- Stop generating some logrus.Fields objects when we don't need them
- Don't look up identities when we already have them.

All in all, this results in a bit faster dns throughput in an artificial test I set up, as well as reduced CPU usage in the critical path. Log enrichment went from ~6% of CPU to around 2.5%, all while requests-per-second increased.

squeed committed:
monitor / hubble: delay converting endpoint LabelArray to string

While stress-testing the FQDN subsystem, I noticed a lot of memory
allocations (and corresponding GC churn) was due to generating Hubble
events.

Given that these fields are largely thrown away, let's delay converting
the LabelArray in to an array of strings until the last possible point.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
proxy: don't generate log fields unless needed, minor cleanups

I noticed that generating these fields consumed a large amount of CPU in
the latency-critical notifyOnDNSMsg function. Given that flow debugging
is usually disabled, put the field generation behind a closure that is
only called when required.

Additionally, generate map fields more efficiently (thanks for the
suggestion, @learnitall).

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
fqdn / proxy: pre-populate some hubble fields we already know

I noticed that a lot of CPU time was spent determining endpoint labels
for hubble logging purposes. This is somewhat silly, since we already
know them.

So, update the EndpointInfoRegistry to stop filling in labels if they're
already present.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

learnitall commented on pkg/envoy/accesslog_server.go:
```diff
@@ -147,8 +147,10 @@ func (s *AccessLogServer) handleConn(ctx context.Context, conn *net.UnixConn) {
-		flowdebug.Log(log.WithFields(logrus.Fields{}),
-			fmt.Sprintf("%s: Access log message: %s", pblog.PolicyName, pblog.String()))
+		flowdebug.Log(func() (*logrus.Entry, string) {
+			return log.WithFields(logrus.Fields{}),
```
I think the call to `WithFields` is avoidable here. The `log` variable is a `*logrus.Entry` itself, so it can be returned as-is. Maybe the original intent was to convert a `*Logger` struct into an `*Entry` struct?

squeed replied:
hah, good catch!

learnitall commented on pkg/proxy/logger/logger.go:
```diff
@@ -220,7 +220,9 @@ func (lr *LogRecord) getLogFields() *logrus.Entry {
 func (lr *LogRecord) Log() {
-	flowdebug.Log(lr.getLogFields(), "Logging flow record")
+	flowdebug.Log(func() (*logrus.Entry, string) {
+		return lr.getLogFields(), "Logging flow record"
```
I was taking a look at the method `getLogFields` and I think it can be optimized a bit if you'd like to roll that into this PR. The method `getLogFields` makes, at worse, two calls to `WithFields` and we could collapse it into one:

```patch
diff --git a/pkg/proxy/logger/logger.go b/pkg/proxy/logger/logger.go
index 9fdd237a79..a31f4cde08 100644
--- a/pkg/proxy/logger/logger.go
+++ b/pkg/proxy/logger/logger.go
@@ -190,32 +190,28 @@ func (lr *LogRecord) ApplyTags(tags ...LogTag) {
 }
 
 func (lr *LogRecord) getLogFields() *logrus.Entry {
-       fields := log.WithFields(logrus.Fields{
+       fields := logrus.Fields{
                FieldType:    lr.Type,
                FieldVerdict: lr.Verdict,
                FieldMessage: lr.Info,
-       })
+       }
 
        if lr.HTTP != nil {
-               fields = fields.WithFields(logrus.Fields{
-                       FieldCode:     lr.HTTP.Code,
-                       FieldMethod:   lr.HTTP.Method,
-                       FieldURL:      lr.HTTP.URL,
-                       FieldProtocol: lr.HTTP.Protocol,
-                       FieldHeader:   lr.HTTP.Headers,
-               })
+               fields[FieldCode] = lr.HTTP.Code
+               fields[FieldMethod] = lr.HTTP.Method
+               fields[FieldURL] = lr.HTTP.URL
+               fields[FieldProtocol] = lr.HTTP.Protocol
+               fields[FieldHeader] = lr.HTTP.Headers
        }
 
        if lr.Kafka != nil {
-               fields = fields.WithFields(logrus.Fields{
-                       FieldCode:               lr.Kafka.ErrorCode,
-                       FieldKafkaAPIKey:        lr.Kafka.APIKey,
-                       FieldKafkaAPIVersion:    lr.Kafka.APIVersion,
-                       FieldKafkaCorrelationID: lr.Kafka.CorrelationID,
-               })
+               fields[FieldCode] = lr.Kafka.ErrorCode
+               fields[FieldKafkaAPIKey] = lr.Kafka.APIKey
+               fields[FieldKafkaAPIVersion] = lr.Kafka.APIVersion
+               fields[FieldKafkaCorrelationID] = lr.Kafka.CorrelationID
        }
 
-       return fields
+       return log.WithFields(fields)
 }
 
 // Log logs a record to the logfile and flushes the buffer
```

kaworu replied:
> The method getLogFields makes, at worse, two calls to WithFields

Correct as `lr.Kafka != nil` and `lr.HTTP != nil` should be mutually exclusive, thus maybe we could refactor in a `switch` statement (or `if` / `else if`) to make it clearer.

Ryan's suggestion also will not trigger a map grow since we're at most setting 8 key/val in the HTTP case, maybe we could explicitly initialize the map with the known max key/val set by `getLogFields`:

```
// NOTE: we're adding at most 8 key/val from this func, allocate enough space to avoid having to grow the map.
fields := make(logrus.Fields, 8)
fields[FieldType] = lr.Type
fields[FieldVerdict] = lr.Verdict
fields[FieldMessage] = lr.Info
```

squeed replied:
Applied both of these suggestions, thanks!

learnitall commented on daemon/cmd/fqdn.go:
```diff
@@ -318,13 +318,15 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
 		addrInfo.DstIPPort = epIPPort
-		addrInfo.DstIdentity = ep.GetIdentity()
+		addrInfo.DstEPID = ep.GetID()
+		addrInfo.DstSecIdentity, _ = ep.GetSecurityIdentity()
```
Would you mind adding a quick comment explaining why it's okay that we can avoid checking the error here? It'll probably save folks some time in the future.

learnitall commented on pkg/proxy/logger/endpoint/epinfo.go:
```diff
@@ -40,42 +40,45 @@ func NewEndpointInfoRegistry(ipc *ipcache.IPCache, endpointManager endpointmanag
 	// reported by the proxy. This way we log the security identity and labels used for
 	// policy enforcement, if any.
-	if id == 0 {
+	if info.ID == 0 || info.Identity == 0 {
```
If `ID` is equal to zero, is `Identity` also always equal to 0? In the case that only one is not set, we'd waste time updating both.

squeed replied:
Good question. The unfortunately-named ID is actually the Endpoint ID, not the security Identity. There are cases in the proxy where we already know the remote identity, but not the endpoint.

Of course, we can't know the endpoint ID if it is remote, but if we proxy a DNS request to coredns on the same node, then it is possible to add the endpoint ID as well.

squeed replied:
However, you are right, the conditions aren't clear. I made the logic nicer.

learnitall reviewed (commented):
Great work! Just a couple of comments and suggestions here. Thank you!

kaworu commented on pkg/proxy/accesslog/record.go:
```diff
@@ -83,7 +85,7 @@ type EndpointInfo struct {
 
 	// Labels is the list of security relevant labels of the endpoint
-	Labels []string
+	Labels labels.LabelArray
```
Before this patch the `EndpointInfo` struct owned this member but after it's a reference to the identity's `LabelArray`. I checked all refs and we're safe but IMO we should add a note here that this member is read-only and should not be mutated.

squeed replied:
fair point. Yes, we should not be mutating labels.

My kingdom for an immutable reference type.

kaworu commented on pkg/proxy/logger/endpoint/epinfo.go:
```diff
@@ -40,42 +40,45 @@ func NewEndpointInfoRegistry(ipc *ipcache.IPCache, endpointManager endpointmanag
-		info.Labels = identity.Labels.GetModel()
+
+	if info.Labels == nil {
+		identity := r.identityAllocator.LookupIdentityByID(context.TODO(), identity.NumericIdentity(info.Identity))
```
I understand is was there before the patch, but are we ok with `context.TODO()` here? Shouldn't we at least use a timeout?

squeed replied:
We should probably add a timeout, yes. This logging is in the critical path for the DNS proxy.

Fortunately, this goes through the identity cache, so the chances of timeout are very, very low. I'll add it as a separate commit.

kaworu reviewed (commented):
Thanks for the patch @squeed! Left a couple of comments / questions but overall LGTM.

squeed committed:
proxy logging: restrict label retrieval time

We resolve security identities while generating a proxied DNS response. This
leaves us open to unbounded latency in the response path. So, cap the
lookup time at 10ms. We have a total latency budget of 100ms, so this is
a reasonable backstop.

Since these identities are almost certainly already in cache, there is
little risk that we lose log fields. Plus, this protects us against DNS tail
latency spikes.

Suggested-by: Alexandre Perrin <alex@isovalent.com>
Signed-off-by: Casey Callendrello <cdc@isovalent.com>

kaworu commented on daemon/cmd/fqdn.go:
```diff
@@ -456,10 +461,15 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
+	//
+	// Restrict label enrichment time to 10ms; we don't want to block DNS
+	// requests because an identity isn't in the local cache yet.
+	logContext, lcncl := context.WithTimeout(d.ctx, 10*time.Millisecond)
```
Given how much lookups are done by `FillEndpointInfo` (called twice by `Addressing`) I'm unsure whether 10ms is a good compromise here.

squeed replied:
What's your thinking here? Too high, or too low? Remember, we have a latency commitment for DNS requests of 100ms, so 10ms isn't totally out of order here.

The context / timeout is only relevant when fetching identity labels from the kvstore. I'm fairly certain this will never actually happen, since the identities should always be in the local cache.

Specifically, this is a flow we intercepted as a result of L7 policy. With the exception of wildcard selectors, these identities must, by definition, already be in cache, since they're in the policy map. One side of the flow is a local pod, so we must have its identity for policy. The other side of the flow is generally the `kube-dns` pod, which should have a stable identity.

So, this timeout is probably irrelevant, just a circuit-breaker to prevent yet another source of dns tail latency.

kaworu replied:
> What's your thinking here? Too high, or too low?

I had in mind the status-quo of no timeout at all, to which 10ms "seems small" and a potential source of flake.

> Remember, we have a latency commitment for DNS requests of 100ms, so 10ms isn't totally out of order here.

Thanks for pointing it out I wasn't aware of that! Sure 10ms sounds reasonable then.

Now, looking down the call stack the log is sent to the monitor agent, which notify/send its observer/listeners. Concretely, only Hubble subscribe as an observer and only the monitor API register as listener (both enabled by default, controlled by `--enable-hubble` and `--enable-monitor`, respectively).

When both are disabled then we're needlessly paying the logging cost in the hot codepath, and for `--enable-monitor` we're synchronously [gob encoding the message](https://github.com/cilium/cilium/blob/acd761befdbd2d13a4e5dc28887e7d887d4b3afc/pkg/monitor/agent/agent.go#L166-L172) even when there is no active listener at the monitor API level (i.e. no ongoing `cilium monitor`).

If correct then there is room for improvement. I feel like logging, encoding, and metrics accounting should be done asynchronously with respect to the hot DNS request codepath. However that goes beyond the scope PR, should we create a new issue about this?

squeed replied:
You're absolutely right, we should move this work out of the critical section, and avoid doing it needlessly where possible. That is a bigger change than I'd like for this PR, though. I know @marseel was looking in to deferring the log enrichment.

As an interesting point, we don't want to defer enrichment for too long, as identities and IPs could theoretically be reused. I'm not sure if that's immediately relevant, but it is a consideration.

tommyp1ckles replied:
This seems like a fair compromise.

However one concern might be that in the kvstore implementation, under some scenarios, we might be placing some extra load on the kvstore server by starting then cancelling a bunch of requests (although, practically, this shouldn't happen because this will eventually find it's way into the local store?).

squeed replied:
Do we even coalesce kvstore requests? If not, then we're sending, basically, the same flood of requests.

learnitall commented on pkg/proxy/logger/endpoint/epinfo.go:
```diff
@@ -59,23 +66,38 @@ func (r *endpointInfoRegistry) FillEndpointInfo(info *accesslog.EndpointInfo, ad
 		if ep != nil {
-			id = ep.GetIdentity()
-		} else if addr.IsValid() {
+			info.ID = ep.GetID()
```
This might be unnecessary. If `ep != nil` here, then `info.ID` should already be set by the above block right?

squeed replied:
oh derp, good point.

learnitall reviewed (approved):
Looks great, thank you! I just left a small non-blocking comment. LGTM!

jrajahalme reviewed (approved):
LGTM

tommyp1ckles commented on pkg/proxy/logger/endpoint/epinfo.go:
```diff
@@ -59,23 +64,35 @@ func (r *endpointInfoRegistry) FillEndpointInfo(info *accesslog.EndpointInfo, ad
 		if ep != nil {
-			id = ep.GetIdentity()
-		} else if addr.IsValid() {
+			secid, _ := ep.GetSecurityIdentity() // ignore error; this is best-effort
```
nit: relying on the returned valued being nil/zero when err != nil is generally but not necessarily true.

squeed replied:
hey man, you got any security IDs? I'm desperate. I'll take anything.


Good point.

tommyp1ckles reviewed (approved):
nice! looks good
